src/inference/ess.jl

```diff
@@ -0,0 +1,134 @@
+"""
+    ESS
+
+Elliptical slice sampling algorithm.
+
+# Examples
+```jldoctest; setup = :(Random.seed!(1))
+julia> @model gdemo(x) = begin
+           m ~ Normal()
+           x ~ Normal(m, 0.5)
+       end
+gdemo (generic function with 2 methods)
+
+julia> sample(gdemo(1.0), ESS(), 1_000) |> mean
+Mean
+
+│ Row │ parameters │ mean     │
+│     │ Symbol     │ Float64  │
+├─────┼────────────┼──────────┤
+│ 1   │ m          │ 0.824853 │
+```
+"""
+struct ESS{space} <: InferenceAlgorithm end
+
+ESS() = ESS{()}()
+ESS(space::Symbol) = ESS{(space,)}()
+
+mutable struct ESSState{V<:VarInfo} <: AbstractSamplerState
+    vi::V
+end
+
+function Sampler(alg::ESS, model::Model, s::Selector)
+    # sanity check
+    vi = VarInfo(model)
+    space = getspace(alg)
+    vns = _getvns(vi, s, Val(space))
+    length(vns) == 1 ||
+        error("[ESS] does only support one variable ($(length(vns)) variables specified)")
+    dist = getdist(vi, vns[1][1])
+    isgaussian(dist) ||
+        error("[ESS] only supports Gaussian prior distributions")
+
+    state = ESSState(vi)
+    info = Dict{Symbol, Any}()
+
+    return Sampler(alg, info, s, state)
+end
+
+isgaussian(dist) = false
+isgaussian(::Normal) = true
+isgaussian(::NormalCanon) = true
+isgaussian(::AbstractMvNormal) = true
+
+# always accept in the first step
+function step!(::AbstractRNG, model::Model, spl::Sampler{<:ESS}, ::Integer; kwargs...)
+    return Transition(spl)
+end
+
+function step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:ESS},
+    ::Integer,
+    ::Transition;
+    kwargs...
+)
+    # obtain mean of distribution
+    vi = spl.state.vi
+    vn = _getvns(vi, spl)[1][1]
+    dist = getdist(vi, vn)
+    μ = vectorize(dist, mean(dist))
+
+    # obtain previous sample
+    f = vi[vn]
+
+    # recompute log-likelihood in logp
+    if spl.selector.tag !== :default
+        runmodel!(model, vi, spl)
+    end
+    setgid!(vi, spl.selector, vn)
+
+    # sample log-likelihood threshold for the next sample
+    threshold = getlogp(vi) - randexp(rng)
+
+    # sample from the prior
+    ν = vectorize(dist, rand(rng, dist))
+
+    # sample initial angle
+    θ = 2 * π * rand(rng)
+    θₘᵢₙ = θ - 2 * π
+    θₘₐₓ = θ
+
+    while true
+        # compute proposal and apply correction for distributions with nonzero mean
+        sinθ, cosθ = sincos(θ)
+        a = 1 - (sinθ + cosθ)
+        vi[vn] = @. f * cosθ + ν * sinθ + μ * a
+
+        # recompute log-likelihood and check if threshold is reached
+        runmodel!(model, vi, spl)
+        if getlogp(vi) > threshold
```

Comment:
I feel it would be good add a max tries condition here, to ensure the while loop will always terminate. You could give a warning if this occurs and return the initial theta.

Reply:
Or adjust the threshold instead of returning the initial theta. Whatever is more sensible.

Reply:
I'm not sure, both approaches feel wrong since they seem to impact the correctness of the sampling algorithm. I think, either one could error if the threshold is not reached within a certain number of iterations or one could loop until the threshold is reached but output a warning after a certain number of iterations.

Reply:
If I understand it correctly, return the initial theta would simply mean you rejected all proposals. This should still be correct. 

Is it guaranteed that the threshold will be reached? I don’t see why that should be the case. Anyways, throwing an error sounds sensible.

Reply:
Disclaimer: I read the paper a while ago, so please ignore my comments if they don’t make sense.

Reply:
I think there should be an error and not a warning, just to make it very obvious that the max iteration limit needs to be adjusted or that there is some numerical issue with the inference (if that actually happens with ESS).

Reply:
> I feel it would be good add a max tries condition here, to ensure the while loop will always terminate.

I shouldn't have commented on that while being on a flight and without checking it more carefully - actually by definition the loop will always terminate. The old theta satisfies the exit condition and hence if the bracket is shrunk enough a theta close/equal to the old theta will be picked as the next theta (but the chain might be stuck and mix slowly). Due to this behaviour ESS can be inefficient (see, e.g., some comments in the original paper and https://arxiv.org/pdf/1210.7477.pdf) but I'm not sure we should address this in our implementation, since the sampling step is guaranteed to terminate. Murray's MATLAB code does not impose a maximum number of iterations, and it feels more like the main issue would be that the chain mixes slowly.

Reply:
I wouldn’t address it if it is guaranteed to terminate. This was my only concern.

Reply:
Thanks for clarifying!